### PR TITLE
max-inflight tuner

### PR DESCRIPTION
**Pull request by tkashem**
This is an MVP implementation of a basic controller that tunes max-inflight values based on the following trigger:
- double the max-inflight values if the number of nodes exceeds a  given threshold
- double the max-inflight values if the number of pods exceeds a given threshold

TODO:
- The Pod and Node threshold are hard coded today, making it configurable is left as a todo for the future if it proves to be useful.

### CODE REVIEW

File: pkg/operator/maxinflighttunercontroller/controller.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+		eventRecorder:  eventRecorder,
+		defaults:       *defaults,
+		metricsClient:  metrics.NewClient(metricsClient),
+	}
+
+	controller = factory.New().ResyncEvery(ResyncInterval).WithSync(syncer.sync).ToController("MaxInFlightTunerController", eventRecorder)
+	return
+}
+
+type MaxInflightTunerController struct {
```
Comment (sttts) [inline pkg/operator/maxinflighttunercontroller/controller.go:64]: please document on a high level what this controller does

File: pkg/operator/maxinflighttunercontroller/maxinflight_controller.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/metrics"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const (
+	// If the total number of nodes exceeds this threshold we will double the values for
+	// max-requests-inflight and max-mutating-requests-inflight.
+	NodeThreshold = 2
```
Comment (tkashem) [inline pkg/operator/maxinflighttunercontroller/maxinflight_controller.go:22]: Set to 2 to test it in CI.

File: pkg/operator/maxinflighttunercontroller/maxinflight_controller.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+// - if the number of Pods exceeds the default threshold (defined by PodThreshold)
+// - if the number of Nodes exceeds the default threshold (defined by NodeThreshold)
+func (m *MaxInflightTunerController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	summary, err := m.metricsClient.GetNodeAndPodSummary(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve metrics - %s", err.Error())
+	}
+	desired := GetDesiredMaxInFlightValues(m.defaults, summary)
+
+	klog.V(3).Infof("desired max-inflight settings:%s metrics summary:%s node threshold:%d pod threshold:%d", desired.String(), summary.String(), NodeThreshold, PodThreshold)
```
Comment (tkashem) [inline pkg/operator/maxinflighttunercontroller/maxinflight_controller.go:88]: I will reset it to `V(5)` when I am done looking in CI.

**Comment by tkashem** [comment]
Need to reset a few hard coded value before it can get in.

/hold

File: pkg/operator/maxinflighttunercontroller/maxinflight_controller.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+)
+
+const (
+	// If the total number of nodes exceeds this threshold we will double the values for
+	// max-requests-inflight and max-mutating-requests-inflight.
+	NodeThreshold = 2
+
+	// If the total number of pods exceeds this threshold we will double the values for
+	// max-requests-inflight and max-mutating-requests-inflight.
+	PodThreshold = 1000
```
Comment (mfojtik) [inline pkg/operator/maxinflighttunercontroller/maxinflight_controller.go:26]: are we consuming these const anywhere? if not make them lowercase :)

File: pkg/operator/maxinflighttunercontroller/maxinflight_controller.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	if !updated {
+		return nil
+	}
+
+	klog.Infof("successfully tuned max-inflight settings new values:%s metrics summary:%s", desired.String(), summary.String())
+	m.eventRecorder.Eventf("MaxInFlightSettingsUpdated", "successfully tuned max-inflight settings new values:%s metrics summary:%s", desired.String(), summary.String())
+	return nil
+}
+
+func GetDesiredMaxInFlightValues(defaults MaxInFlightValues, summary *metrics.Summary) MaxInFlightValues {
```
Comment (mfojtik) [inline pkg/operator/maxinflighttunercontroller/maxinflight_controller.go:107]: private

File: pkg/operator/maxinflighttunercontroller/maxinflightreadwriter.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+const (
+	MaxMutatingRequestsInFlight = "max-mutating-requests-inflight"
+	MaxReadOnlyRequestsInFlight = "max-requests-inflight"
+	AssetName                   = "v4.1.0/config/defaultconfig.yaml"
+)
+
+type MaxInFlightValues struct {
+	// MaxReadOnlyInflight holds the value of 'max-requests-inflight'.
+	// A nil means no value has been specified.
+	MaxReadOnlyInflight *int
```
Comment (mfojtik) [inline pkg/operator/maxinflighttunercontroller/maxinflightreadwriter.go:24]: nit MaxReadOnlyInFlight

File: pkg/operator/maxinflighttunercontroller/maxinflight_controller.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	// max-mutating-requests-inflight values if necessary.
+	ResyncInterval = 1 * time.Hour
+
+	// Name of the controller.
+	ControllerName = "MaxInFlightTunerController"
+)
+
+// NewMaxInflightTunerController returns a new instance of the controller.
+func NewMaxInflightTunerController(config *rest.Config, operatorClient operatorv1helpers.StaticPodOperatorClient, eventRecorder events.Recorder) (factory.Controller, error) {
+	metricsClient, err := metricsclient.NewForConfig(config)
```
Comment (mfojtik) [inline pkg/operator/maxinflighttunercontroller/maxinflight_controller.go:38]: do this outside, pass the client here from starter

File: pkg/operator/maxinflighttunercontroller/maxinflight_controller.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+		klog.Infof("[max-inflight-tuner] no tuning required - %s", values)
+		return nil
+	}
+	_, specUpdated, err := operatorv1helpers.UpdateSpec(m.operatorClient, func(spec *operatorv1.OperatorSpec) error {
+		return WriteMaxInFlightValues(spec, desiredMaxInFlightValues)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to tune max-inflight arguments - %s - %s", values, err.Error())
+	}
+	if !specUpdated {
```
Comment (vareti) [inline pkg/operator/maxinflighttunercontroller/maxinflight_controller.go:111]: is this check required?

**Comment by vareti** [comment]
actually never mind. i understand now how scale down works. Please ignore my previous question.

**Pull request closed without merging**